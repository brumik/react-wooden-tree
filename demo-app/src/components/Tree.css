li {
    list-style: none;
}
.NoOpenButton {
    margin-left: 18px;
}
.Icon {
    width: 14px;
    margin: 2px;
}
.NodeIconImage {
    width: 14px;
    height: auto;
    margin: 0 4px -1px;
}
.Selectable {
    cursor: pointer;
}