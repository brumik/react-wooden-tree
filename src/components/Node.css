.indent-1 {
    padding-left: 15px
}
.indent-2 {
    padding-left: 30px
}
.indent-3 {
    padding-left: 45px
}
.indent-4 {
    padding-left: 60px
}
.indent-5 {
    padding-left: 75px
}
