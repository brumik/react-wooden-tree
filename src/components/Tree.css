li {
    list-style: none;
}
.NoOpenButton {
    margin-left: 18px;
}
.SelectButton {
    width: 14px;
    background: transparent;
    border: none;
    outline: none;
    margin: 2px;
    padding: 0;
    cursor: pointer;
}
.ExpandButton {
    width: 14px;
    background: transparent;
    border: none;
    outline: none;
    margin: 2px;
    padding: 0;
}
.NodeIconImage {
    width: 14px;
    height: auto;
    margin: 0 4px -1px;
}