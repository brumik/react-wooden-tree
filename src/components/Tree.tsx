import * as React from "react";
import { ListProps, List } from "./List";

export interface TreeProps {
    id: string
    list?: ListProps
}

interface TreeState {
}

export class Tree extends React.Component<TreeProps, TreeState> {
    constructor(props: TreeProps) {
        super(props);
    }

    render() {
        return (
            <div id={this.props.id}>
                <p>Will be a tree</p>
                <List
                    id={this.props.id}
                    {...this.props.list}
                />
            </div>
        );
    }
}