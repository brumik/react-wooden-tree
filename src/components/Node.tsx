import * as React from 'react';
import { NodeProps, ParentDataType } from '..';
import { CheckboxButton } from './CheckboxButton';
import { ExpandButton } from './ExpandButton';
import { ParentDataContext } from './Tree';

/**
 * @class Node
 * @extends React.Component
 *
 * Displays a node and communicates with submodules and tree.
 */
export class Node extends React.PureComponent<NodeProps, {}> {
    /**
     * Used for default values.
     */
    public static defaultProps: NodeProps;

    /**
     * Used for determining if checkbox has changed.
     */
    private readonly defaultCheckbox: boolean;

    /**
     * Creates the Node[] components from given nodes.
     *
     * @param {string[]} nodeIds The nodes to render.
     * @param {ParentDataType} parentData The parent data to pass.
     * @returns {JSX.Element[]} The array of JSX elements with nodes.
     */
    public static renderSublist(nodeIds: string[], { connectedNode, tree }: ParentDataType): JSX.Element[] {
        let elements: JSX.Element[] = [];
        if ( connectedNode ) {
            const ConnectedNode = connectedNode;
            for (let i = 0; i < nodeIds.length; i++) {
                elements.push(
                    <ConnectedNode
                        key={nodeIds[i]}
                        nodeId={nodeIds[i]}
                    />
                );
            }
        } else {
            for (let i = 0; i < nodeIds.length; i++) {
                elements.push(
                    <Node
                        key={nodeIds[i]}
                        nodeId={nodeIds[i]}
                        {...tree[nodeIds[i]]}
                    />
                );
            }
        }
        return elements;
    }

    /**
     * Renders the tree element.
     *
     * @returns {JSX.Element}
     */
    public render (): JSX.Element {
        // Indent class
        let NodeClasses = 'indent-' + this.getItemIndentSize();

        // Checkbox
        const checkbox = !this.props.hideCheckbox && this.context.showCheckbox ? (
            <CheckboxButton
                onChange={this.handleCheckChange}
                checked={this.props.state.checked}
                checkedIcon={this.context.checkedIcon}
                partiallyCheckedIcon={this.context.partiallyCheckedIcon}
                uncheckedIcon={this.context.uncheckedIcon}
            />
        ) : null;

        // Expand button: if not displayed added padding
        let openButton: JSX.Element;
        if ( ( this.props.nodes && this.props.nodes.length > 0 ) || this.props.lazyLoad ) {
            openButton = (
                <ExpandButton
                    onChange={this.handleOpenChange}
                    expanded={this.props.state.expanded}
                    loading={this.props.loading}
                    expandIcon={this.context.expandIcon}
                    collapseIcon={this.context.collapseIcon}
                    loadingIcon={this.context.loadingIcon}
                    errorIcon={this.context.errorIcon}
                />
            );
        } else {
            openButton = null;
            NodeClasses += ' no-open-button';
        }

        // Icon
        let icon: JSX.Element = null;
        if ( this.context.showIcon ) {

            let iconStyle: {color?: string, backgroundColor?: string} = {};

            if ( this.props.iconColor ) {
                iconStyle.color = this.props.iconColor;
            }

            if ( this.props.iconBackground ) {
                iconStyle.backgroundColor = this.props.iconBackground;
            }

            if ( this.context.showImage && this.props.image ) {
                icon = <img className={'icon image'} src={this.props.image}/>;
            } else if ( this.props.icon ) {
                icon = <i className={'icon ' + this.props.icon} style={iconStyle}/>;
            } else {
                icon = <i className={'icon ' + this.context.nodeIcon} style={iconStyle}/>;
            }
        }

        // Selected
        let selectedIcon: JSX.Element = null;
        if ( this.props.state.selected ) {
            if ( this.props.selectedIcon ) {
                selectedIcon = <i className={this.props.selectedIcon}/>;
            } else {
                selectedIcon = <i className={this.context.selectedIcon}/>;
            }
        }

        // Selectable class
        if ( this.props.selectable && !this.props.state.disabled ) {
            NodeClasses += ' selectable';
        }

        // Children
        const sublist = this.props.state.expanded ? (
            Node.renderSublist(this.props.nodes, this.context)
        ) : null;

        // Determining icon order
        let icon1, icon2: JSX.Element;
        if ( this.context.checkboxFirst ) {
            icon1 = checkbox;
            icon2 = icon;
        } else {
            icon1 = icon;
            icon2 = checkbox;
        }

        // Additional classes
        if ( this.props.classes ) {
            NodeClasses += ' ' + this.props.classes;
        }
        // Changed checkbox class
        if ( this.props.state.checked !== this.defaultCheckbox ) {
            NodeClasses += ' ' + this.context.changedCheckboxClass;
        }
        // Selected class
        if ( this.props.state.selected ) {
            NodeClasses += ' ' + this.context.selectedClass;
        }

        // Data Attributes list
        let DataAttr: {[k: string]: string} = {'data-id': this.props.nodeId};

        return (
            <React.Fragment>
                <li className={NodeClasses} {...DataAttr} {...this.props.attr}>
                    {openButton}
                    {icon1}
                    {selectedIcon}
                    {icon2}
                    <span onClick={this.handleSelected} dangerouslySetInnerHTML={{ __html: this.props.text }} />
                </li>
                {sublist}
            </React.Fragment>
        );
    }

    /**
     * Constructor.
     * @param {NodeProps} props
     */
     public constructor(props: NodeProps, context: ParentDataType) {
        super(props, context);

        if ( this.props.state.selected ) {
            this.context.initSelectedNode(this.props.nodeId);
        }

        this.defaultCheckbox = this.props.state.checked;
    }

    /**
     * Own checkbox handler.
     * @param {boolean} checked Contains the input field value.
     */
    private handleCheckChange = (checked: boolean): void => {
        if ( this.props.checkable && !this.props.state.disabled ) {
            this.context.checkboxOnChange(checked, this.props.nodeId);
        }
    }

    /**
     * Handles open event.
     * If lazy load set then loads the child nodes too.
     *
     * @param {boolean} expanded True on expand false on collapse.
     */
    private handleOpenChange = (expanded: boolean): void => {
        if ( this.props.lazyLoad && this.props.nodes.length < 1 ) {
            this.context.onLazyLoad(this.props.nodeId);
        } else {
            this.context.expandOnChange(this.props.nodeId, expanded);
        }
    }

    /**
     * Handles the selected event. If allowed to select then calls the callback
     * function with the opposite of currently selected state.
     *
     * If select is disabled on the node then it works like expand.
     */
    private handleSelected = (): void => {
        if ( this.props.selectable && !this.props.state.disabled ) {
            this.context.selectOnChange(this.props.nodeId, !this.props.state.selected);
        } else {
            this.handleOpenChange(!this.props.state.expanded);
        }
    }

    /**
     * Returns the computed padding size for the current list item for indent.
     * @returns {number} The computed padding level.
     */
    private getItemIndentSize = (): number => this.props.nodeId.split('.').length - 1;
}

Node.contextType = ParentDataContext;

/**
 * Node default values.
 */
Node.defaultProps = {
    nodeId: '',
    text: '',
    nodes: [],
    state: {
        checked: false,
        expanded: false,
        disabled: false,
        selected: false
    },

    checkable: true,
    hideCheckbox: false,
    selectable: true,
    selectedIcon: null,
    lazyLoad: false,
    loading: false,

    attr: null,

    // Styling
    icon: null,
    iconColor: null,
    iconBackground: null,
    image: null,
    classes: '',
};
