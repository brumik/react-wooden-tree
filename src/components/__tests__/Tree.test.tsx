import { Tree } from '../Tree';
import { NodeProps } from '../Node';

let tree: NodeProps[];
let subTree: NodeProps[];
const initState = {
    checked: false,
    selected: false,
    disabled: false,
    expanded: false
};

beforeEach(() => {
    tree = [
        {text: 'Parent 0',
            nodes: [
                {text: 'Child 0.0'},
                {text: 'Child 0.1',
                    nodes: [
                        {text: 'Child 0.1.0'}
                    ]},
                {text: 'Child 0.2'}
            ]},
        {text: 'Parent 1', state: {selected: true}}
    ];

    subTree = [
        {text: 'Sub Parent 0'},
        {text: 'Sub Parent 1',
            nodes: [
                {text: 'Sub Child 1.0'},
                {text: 'Sub Child 1.1'}
            ]},
        {text: 'Sub Parent 2',
            nodes: [
                {text: 'Sub Child 2.0', state: {checked: true}}
            ]},
    ];

    tree = Tree.initTree(tree);
});

describe('tree public method', () => {
    it('should initialize the ids and state correctly', () => {
        expect(tree).not.toBeNull();

        expect(tree[0].nodes[1].nodes[0].id).toBe('0.1.0');
        expect(tree[0].nodes[1].nodes[0].state).toMatchObject(initState);

        expect(tree[0].nodes[0].id).toBe('0.0');
        expect(tree[0].nodes[0].state).toMatchObject(initState);

        expect(tree[1].id).toBe('1');
        expect(tree[1].state).toMatchObject({...initState, selected: true});
    });

    it('should return the correct node', () => {
        let node = Tree.nodeSelector(tree, '0.2');
        expect(node.text).toBe('Child 0.2');

        node = Tree.nodeSelector(tree, '1');
        expect(node.text).toBe('Parent 1');
    });

    it('should update the tree correctly', () => {
        let node = Tree.nodeSelector(tree, '1');

        node.text = 'Edited Node';
        Tree.nodeUpdater(tree, node);
        expect(tree[1].text).toBe('Edited Node');

        node = Tree.nodeSelector(tree, '0.1.0');
        node.text = 'Edited Child Node';
        expect(tree[0].nodes[1].nodes[0].text).toBe('Edited Child Node');
    });

    it('should return node with changed checked state', () => {
        let node = Tree.nodeSelector(tree, '1');

        node = Tree.nodeChecked(node, true);
        expect(node.state).toMatchObject({
            checked: true,
            expanded: false,
            disabled: false,
            selected: true
        });

        node = Tree.nodeChecked(node, false);
        expect(node.state).toMatchObject({
            checked: false,
            expanded: false,
            disabled: false,
            selected: true
        });
    });

    it('should return node with changed expanded state', () => {
        let node = Tree.nodeSelector(tree, '1');

        node = Tree.nodeExpanded(node, true);
        expect(node.state).toMatchObject({
            expanded: true,
            disabled: false,
            selected: true,
            checked: false
        });

        node = Tree.nodeExpanded(node, false);
        expect(node.state).toMatchObject({
            expanded: false,
            disabled: false,
            selected: true,
            checked: false
        })
    });

    it('should return node with changed disabled state', () => {
        let node = Tree.nodeSelector(tree, '1');

        node = Tree.nodeDisabled(node, true);
        expect(node.state).toMatchObject({
            expanded: false,
            disabled: true,
            selected: true,
            checked: false
        });

        node = Tree.nodeDisabled(node, false);
        expect(node.state).toMatchObject({
            expanded: false,
            disabled: false,
            selected: true,
            checked: false
        })
    });

    it('should return node with changed selected state', () => {
        let node = Tree.nodeSelector(tree, '1');

        node = Tree.nodeSelected(node, false);
        expect(node.state).toMatchObject({
            expanded: false,
            disabled: false,
            selected: false,
            checked: false
        });

        node = Tree.nodeSelected(node, true);
        expect(node.state).toMatchObject({
            expanded: false,
            disabled: false,
            selected: true,
            checked: false
        })
    });

    it('should return node with changed children nodes', () => {
        let node = Tree.nodeSelector(tree, '1');

        subTree = Tree.initTree(subTree, node.id);
        node = Tree.nodeChildren(node, subTree);

        expect(node.nodes).not.toBeNull();
        expect(node.nodes[1].nodes[0]).toMatchObject({
            text: 'Sub Child 1.0', id: '1.1.0', state: initState
        });

        expect(node.nodes[2].nodes[0]).toMatchObject({
            text: 'Sub Child 2.0', id: '1.2.0', state: {...initState, checked: true}
        });
    });

    it('should return node with changed loading state', () => {
        let node = Tree.nodeSelector(tree, '1');

        node = Tree.nodeLoading(node, true);
        expect(node.loading).toBeTruthy();

        node = Tree.nodeLoading(node, false);
        expect(node.loading).toBeFalsy();
    });
});