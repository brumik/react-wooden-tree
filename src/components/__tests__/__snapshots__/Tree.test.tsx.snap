// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tree events should check parent if all children checked 1`] = `
<div
  className="Tree"
>
  <ul>
    <li
      className="indent-0"
      id="0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 0
      </span>
    </li>
    <li
      className="indent-0 Selectable changed-checkbox"
      id="1"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 1
      </span>
    </li>
    <li
      className="indent-1 Selectable changed-checkbox"
      id="1.0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable changed-checkbox selected"
      id="1.0.0"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-check"
      />
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable changed-checkbox"
      id="1.0.1"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.1
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable changed-checkbox"
      id="1.0.2"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.2
      </span>
    </li>
    <li
      className="indent-0 Selectable"
      id="2"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 2
      </span>
    </li>
  </ul>
  <style>
    .indent-1{padding-left:18px}.indent-2{padding-left:36px}.indent-3{padding-left:54px}
  </style>
</div>
`;

exports[`tree events should match failed lazy load 1`] = `
<div
  className="Tree"
>
  <ul>
    <li
      className="indent-0"
      id="0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 0
      </span>
    </li>
    <li
      className="indent-0 Selectable"
      id="1"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 1
      </span>
    </li>
    <li
      className="indent-1 Selectable"
      id="1.0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable selected"
      id="1.0.0"
    >
      <i
        className="fa fa-check"
      />
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable"
      id="1.0.1"
    >
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.1
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable"
      id="1.0.2"
    >
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.2
      </span>
    </li>
    <li
      className="indent-0 Selectable"
      id="2"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa-exclamation-triangle"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 2
      </span>
    </li>
  </ul>
  <style>
    .indent-1{padding-left:18px}.indent-2{padding-left:36px}.indent-3{padding-left:54px}
  </style>
</div>
`;

exports[`tree events should match lazy loaded nodes 1`] = `
<div
  className="Tree"
>
  <ul>
    <li
      className="indent-0"
      id="0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 0
      </span>
    </li>
    <li
      className="indent-0 Selectable"
      id="1"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 1
      </span>
    </li>
    <li
      className="indent-1 Selectable"
      id="1.0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable selected"
      id="1.0.0"
    >
      <i
        className="fa fa-check"
      />
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable"
      id="1.0.1"
    >
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.1
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable"
      id="1.0.2"
    >
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.2
      </span>
    </li>
    <li
      className="indent-0 Selectable"
      id="2"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 2
      </span>
    </li>
    <li
      className="indent-1 NoOpenButton Selectable"
      id="2.0"
    >
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Lazy Loaded
      </span>
    </li>
  </ul>
  <style>
    .indent-1{padding-left:18px}.indent-2{padding-left:36px}.indent-3{padding-left:54px}
  </style>
</div>
`;

exports[`tree events should partially check parent if not all children checked 1`] = `
<div
  className="Tree"
>
  <ul>
    <li
      className="indent-0"
      id="0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 0
      </span>
    </li>
    <li
      className="indent-0 Selectable changed-checkbox"
      id="1"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 1
      </span>
    </li>
    <li
      className="indent-1 Selectable changed-checkbox"
      id="1.0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable changed-checkbox selected"
      id="1.0.0"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-check"
      />
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable"
      id="1.0.1"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.1
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable"
      id="1.0.2"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.2
      </span>
    </li>
    <li
      className="indent-0 Selectable"
      id="2"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 2
      </span>
    </li>
  </ul>
  <style>
    .indent-1{padding-left:18px}.indent-2{padding-left:36px}.indent-3{padding-left:54px}
  </style>
</div>
`;

exports[`tree events should partially check parent if not all children checked 2`] = `
<div
  className="Tree"
>
  <ul>
    <li
      className="indent-0"
      id="0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 0
      </span>
    </li>
    <li
      className="indent-0 Selectable changed-checkbox"
      id="1"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 1
      </span>
    </li>
    <li
      className="indent-1 Selectable changed-checkbox"
      id="1.0"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable changed-checkbox selected"
      id="1.0.0"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-check"
      />
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.0
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable changed-checkbox"
      id="1.0.1"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-check-square"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.1
      </span>
    </li>
    <li
      className="indent-2 NoOpenButton Selectable"
      id="1.0.2"
    >
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Child 1.0.2
      </span>
    </li>
    <li
      className="indent-0 Selectable"
      id="2"
    >
      <button
        className="ExpandButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-right"
        />
      </button>
      <button
        className="SelectButton"
        onClick={[Function]}
      >
        <i
          className="fa fa-square-o"
        />
      </button>
      <i
        className="fa fa-ban fa-fw"
      />
      <span
        onClick={[Function]}
      >
        Parent 2
      </span>
    </li>
  </ul>
  <style>
    .indent-1{padding-left:18px}.indent-2{padding-left:36px}.indent-3{padding-left:54px}
  </style>
</div>
`;
